File: code/852-Peak Index in a Mountain Array/852-Peak Index in a Mountain Array.cpp
/*
Let's call an array arr a mountain if the following properties hold:

arr.length >= 3
There exists some i with 0 < i < arr.length - 1 such that:
arr[0] < arr[1] < ... arr[i-1] < arr[i]
arr[i] > arr[i+1] > ... > arr[arr.length - 1]
Given an integer array arr that is guaranteed to be a mountain, return any i such that arr[0] < arr[1] < ... arr[i - 1] < arr[i] > arr[i + 1] > ... > arr[arr.length - 1].



Example 1:

Input: arr = [0,1,0]
Output: 1
Example 2:

Input: arr = [0,2,1,0]
Output: 1
Example 3:

Input: arr = [0,10,5,2]
Output: 1
Example 4:

Input: arr = [3,4,5,1]
Output: 2
Example 5:

Input: arr = [24,69,100,99,79,78,67,36,26,19]
Output: 2


Constraints:

3 <= arr.length <= 104
0 <= arr[i] <= 106
arr is guaranteed to be a mountain array.


Follow up: Finding the O(n) is straightforward, could you find an O(log(n)) solution?
*/

#include <vector>

using namespace std;

class Solution {
public:
    int peakIndexInMountainArray(vector<int>& arr) {
        int n = arr.size();
        int head = 1, tail = n - 2, ans = 0;
        while (head <= tail) {
            int mid = (head + tail) / 2;
            if (arr[mid] > arr[mid + 1]) {
                ans = mid;
                tail = mid - 1;
            } else {
                head = mid + 1;
            }
        }
        return ans;
    }
};

File: code/852-Peak Index in a Mountain Array/852-Peak Index in a Mountain Array.py
"""
LeetCode 852 - Peak Index in a Mountain Array

解题步骤（自动整理）
1. 先把题目目标拆成可操作的子任务（比较/统计/构造等）
2. 选择合适的数据结构并按一次遍历或分治步骤实现核心逻辑
3. 补齐边界条件（空输入、单元素、重复元素等）并返回结果
"""
# Let's call an array A a mountain if the following properties hold:

# - A.length >= 3
# - There exists some 0 < i < A.length - 1 such that A[0] < A[1] < ... A[i-1] < A[i] > A[i+1] > ... > A[A.length - 1]
# Given an array that is definitely a mountain, return any i such that A[0] < A[1] < ... A[i-1] < A[i] > A[i+1] > ... > A[A.length - 1].

# Example 1:

# Input: [0,1,0]
# Output: 1
# Example 2:

# Input: [0,2,1,0]
# Output: 1
# Note:

# - 3 <= A.length <= 10000
# - 0 <= A[i] <= 10^6
# - A is a mountain, as defined above.

class Solution(object):
    def peakIndexInMountainArray(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        first = 0
        second = len(A)
        if len(A) < 3:
            return None
        for i in range(1, len(A)):
            if A[i]  <= A[i - 1]:
                first = i - 1
                break
        for i in range(len(A) - 2, -1 , -1):
            if A[i] <= A[i + 1]:
                second = i + 1
                break
        return first if first == second else None
